numpy/core/src/multiarray/arrayobject.c

```diff
@@ -475,22 +475,10 @@ array_dealloc(PyArrayObject *self)
                 "called.";
             /* 2017-Nov-10 1.14 */
             if (DEPRECATE(msg) < 0) {
-                /* dealloc must not raise an error, best effort try to write
+                /* dealloc cannot raise an error, best effort try to write
                    to stderr and clear the error
```

Comment:
I think you might need to actually clear the error before calling `PyArray_ResolveWritebackIfCopy`, in case the array is an object dtype, and `PyArray_ResolveWritebackIfCopy` causes refcounts to drop to zero and an error.

Reply:
Although I suppose that's an unrelated fix

Reply:
In the implementation of ``PyErr_WriteUnraisable`` it calls ``PyErr_Fetch`` which clears the error state